typeset/converter.js

```diff
@@ -129,73 +146,44 @@ const createMapOfMathMLElements = async (log, inputPath, cssPath, outputPath, ou
+  })
 
-      // Print progress every 10 seconds
-      const now = Date.now()
-      if (now - prevTime > PROGRESS_TIME) {
-        const percent = Math.round(100 * index / total)
-        console.info(`Inserted ${percent}% of all elements...`)
-        prevTime = now
-      }
-      index++
+  browser.close()
+
+  const BATCH_SIZE = 3000
+  let allUniqueCss = new Set();
+  let nextMathNodeIndex
+  for (let batch = 0; batch < Math.ceil(mathEntries.length / BATCH_SIZE); batch++) {
+    const start = BATCH_SIZE * batch
+    const end = Math.min(BATCH_SIZE * batch + BATCH_SIZE, mathEntries.length)
+    log.info(`Converting math elements ${start} to ${end} of ${mathEntries.length}`)
+    const [convertedMathML/*: Map<string, {svg || html}> */, uniqueCss] = await mjnodeConverter.convertMathML(log, new Map(mathEntries.slice(start, end)), outputFormat, mathEntries.length, start)
+    allUniqueCss.add(uniqueCss)
+
+    log.debug(`Inserting converted math elements...`)
+    let convertedContent = []
+    for (const [id, xml] of convertedMathML) {
+      // Replacing placeholder with converted math
+      nextMathNodeIndex = pageContent.indexOf(mathNodePlaceholder(id))
+      convertedContent.push(pageContent.substr(0, nextMathNodeIndex))
+      convertedContent.push(xml)
+      pageContent = pageContent.substr(nextMathNodeIndex + mathNodePlaceholder(id).length)
     }
-  }, mathSources, PROGRESS_TIME)
-
-  // Inject any CSS that was generated by mathjax-node
-  // The CSS content may be duplicated. If so, remove duplicates.
-  const allCssMaybeDuplicate = [...convertedMathML.values()]
-    .map(({css}) => css)
-    .filter(css => !!css) // only keep the items that have a css block (this is null for svg)
+    await writeFile(output, convertedContent.join(''), {flag: batch === 0 ? 'w' : 'a'})
+  }
+  await writeFile(output, pageContent, {flag: 'a'})
 
-  const allUniqueCss = new Set(allCssMaybeDuplicate)
   log.info(`Injecting MathJax-created CSS...`)
-  await page.evaluate((allCss) => {
-    console.info('Adding MathJax-created CSS')
-    const head = document.querySelector('head')
-    const style = document.createElement('style')
-    style.innerHTML = allCss.join('\n')
-    head.appendChild(style)
-  }, [...allUniqueCss.values()])
+  pageContent = await readFile(output, 'utf-8')
+  await writeFile(output, pageContent.replace('</head>', `<style>${[...allUniqueCss.keys()].join('\n')}</style></head>`, 1))
 
-  log.info(`Serializing XHTML back out...`)
-  let convertedContent = await page.evaluate(() => {
-    console.log('Serializing content...')
-    const s = new window.XMLSerializer()
-    const convertedContent = s.serializeToString(document)
-    return convertedContent
-  })
-
-  log.info('Saving file with injected math HTML elements...')
-  log.debug(`Saving result to "${output}"`)
-  await writeFile(output, convertedContent)
   log.info(`Content saved. Open "${output}" to see converted file.`)
 
-  await browser.close()
```

Comment:
Why is the browser no longer being closed?

Reply:
@philschatz Still being closed. See https://github.com/openstax/mathify/pull/51/files#diff-f3932dbedd8de4dbca5414e981dddabdR156

Reply:
We're not using puppeteer after getting all the math elements so I just closed the browser right after that.